### PR TITLE
Add function twig *_name for intl

### PR DESCRIPTION
**Pull request by seb-jean**
Hi,

I add `language_names`, `script_names`, `country_names`, `locale_names`, `currency_names`, `timezone_names`.

It allows us to have a list but on the Twig side and to use them in filters for example.

### CODE REVIEW

**Comment by seb-jean** [comment]
issue #3741

File: extra/intl-extra/IntlExtension.php [hunk 1]
```diff
@@ -241,6 +247,36 @@ public function getCountryTimezones(string $country): array
             return [];
         }
     }
+    
+    public function getLanguageNames(string $locale = null): array
+    {
+        return Languages::getNames($locale);
```
Comment (stof) [inline extra/intl-extra/IntlExtension.php]: This must handle the MissingResourceException as done in filters

**Review by stof — CHANGES_REQUESTED** [review]

File: extra/intl-extra/IntlExtension.php (hunk 1 above)
Comment (seb-jean) [reply extra/intl-extra/IntlExtension.php]: For instance?
```php
public function getLanguageNames(string $locale = null): array
{
    try {
        return Languages::getNames($locale);
    } catch (MissingResourceException $exception) {
        return [];
    }
}
```

Comment (stof) [reply extra/intl-extra/IntlExtension.php]: yes

**Comment by fabpot** [comment]
Can you add the related docs?

**Comment by seb-jean** [comment]
> Can you add the related docs?

I just added them.

**Comment by SpacePossum** [comment]
maybe some utest (coverage) could be added for the new code?

**Comment by seb-jean** [comment]
> maybe some utest (coverage) could be added for the new code?

I don't know how to do it.

**Comment by GromNaN** [comment]
> > maybe some utest (coverage) could be added for the new code?
> 
> I don't know how to do it.

You can add test files in `extra/intl-extra/Tests/Fixtures` and run the test command:

```
cd extra/intl-extra
composer install
vendor/bin/simple-phpunit
```

**Comment by seb-jean** [comment]
> > > maybe some utest (coverage) could be added for the new code?
> > 
> > 
> > I don't know how to do it.
> 
> You can add test files in `extra/intl-extra/Tests/Fixtures` and run the test command:
> 
> ```
> cd extra/intl-extra
> composer install
> vendor/bin/simple-phpunit
> ```

I just added them.

File: doc/functions/currency_names.rst [hunk 2]
```diff
@@ -0,0 +1,38 @@
+``currency_names``
+=====================
+
+The ``currency_names`` function returns the names of the languages:
```
Comment (GromNaN) [inline doc/functions/currency_names.rst]:
```suggestion
The ``currency_names`` function returns the names of the currencies:
```

File: doc/functions/locale_names.rst [hunk 3]
```diff
@@ -0,0 +1,38 @@
+``locale_names``
+=====================
+
+The ``locale_names`` function returns the names of the languages:
```
Comment (GromNaN) [inline doc/functions/locale_names.rst]:
```suggestion
The ``locale_names`` function returns the names of the locales:
```

File: doc/functions/timezone_names.rst [hunk 4]
```diff
@@ -0,0 +1,38 @@
+``timezone_names``
+=====================
+
+The ``timezone_names`` function returns the names of the languages:
```
Comment (GromNaN) [inline doc/functions/timezone_names.rst]:
```suggestion
The ``timezone_names`` function returns the names of the timezones:
```

**Review by GromNaN — CHANGES_REQUESTED** [review]
Good, the tests are exhaustive.
There is some words to replace in the doc.
The fabbot issue is wrong.

**Review by GromNaN — APPROVED** [review]

**Comment by fabpot** [comment]
It looks like tests are broken.

**Comment by seb-jean** [comment]
> The fabbot issue is wrong.

You are right but I did not touch this part of the code.
fabpot.io offers me the following code:

```php
     public function formatDateTime(Environment $env, $date, ?string $dateFormat = 'medium', ?string $timeFormat = 'medium', string $pattern = '', $timezone = null, string $calendar = 'gregorian', string $locale = null): string
     {
-        $date = \twig_date_converter($env, $date, $timezone);
+        $date = twig_date_converter($env, $date, $timezone);
         $formatter = $this->createDateFormatter($locale, $dateFormat, $timeFormat, $pattern, $date->getTimezone(), $calendar);
 
         if (false === $ret = $formatter->format($date)) 
```

Otherwise, this is the link: https://fabbot.io/patch/twigphp/Twig/3742/cf584f042bf16324db8b8a7a74e067d2e41f03e8/cs.diff

**Comment by seb-jean** [comment]
> It looks like tests are broken.

Tests fail due to line break (\r, \r\n)
I don't see how I could do better to pass the tests.

File: doc/functions/country_names.rst [hunk 5]
```diff
@@ -0,0 +1,38 @@
+``country_names``
+=====================
```
Comment (fabpot) [inline doc/functions/country_names.rst]: The number of `=` must match the title length (same for the other files)

File: doc/functions/country_names.rst [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    .. code-block:: bash
+
+        $ composer require twig/extra-bundle
+
+    Otherwise, add the extension explicitly on the Twig environment::
+
+        use Twig\Extra\Intl\IntlExtension;
+
+        $twig = new \Twig\Environment(...);
+        $twig->addExtension(new IntlExtension());
```
Comment (fabpot) [inline doc/functions/country_names.rst:42]: Can you add an `Arguments` section like for other functions on all doc files? Something like:

```

Arguments
---------

* ``locale``: The locale
```

File: extra/intl-extra/Tests/Fixtures/country_names.test [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
...u, Nepal, Netherlands, New Caledonia, New Zealand, Nicaragua, Niger, Nigeria, Niue, Norfolk Island, North Korea, North Macedonia, Northern Mariana Islands, Norway, Oman, Pakistan, Palau, Palestinian Territories, Panama, Papua New Guinea, Paraguay, Peru, Philippines, Pitcairn Islands, Poland, Portugal, Puerto Rico, Qatar, Réunion, Romania, Russia, Rwanda, Samoa, San Marino, São Tomé & Príncipe, Saudi Arabia, Senegal, Serbia, Seychelles, Sierra Leone, Singapore, Sint Maarten, Slovakia, Slovenia, Solomon Islands, Somalia, South Africa, South Georgia & South Sandwich Islands, South Korea, South Sudan, Spain, Sri Lanka, St. Barthélemy, St. Helena, St. Kitts & Nevis, St. Lucia, St. Martin, St. Pierre & Miquelon, St. Vincent & Grenadines, Sudan, Suriname, Svalbard & Jan Mayen, Sweden, Switzerland, Syria, Taiwan, Tajikistan, Tanzania, Thailand, Timor-Leste, Togo, Tokelau, Tonga, Trinidad & Tobago, Tunisia, Turkey, Turkmenistan, Turks & Caicos Islands, Tuvalu, U.S. Outlying Islands, U.S. Virgin Islands, Uganda, Ukraine, United Arab Emirates, United Kingdom, United States, Uruguay, Uzbekistan, Vanuatu, Vatican City, Venezuela, Vietnam, Wallis & Futuna, Western Sahara, Yemen, Zambia, Zimbabwe
```
Comment (fabpot) [inline extra/intl-extra/Tests/Fixtures/country_names.test]: These tests are going to break every time there is an update in the underlying dataset. I would suggest only listing some "stable" entries instead.

**Review by fabpot — CHANGES_REQUESTED** [review]
Looks good to me.
I made some minor comments.

File: doc/functions/country_names.rst [hunk 8]
```diff
@@ -0,0 +1,43 @@
+``country_names``
+=================
+
```
Comment (fabpot) [inline doc/functions/country_names.rst:3]: You should add something like this on all doc pages:

```
.. versionadded:: 3.5

    The ``country_names`` function was added in Twig 3.5.
```

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by fabpot** [comment]
@seb-jean Thank you fo all the changes. I'm going to take over the tests changes before merging the PR (I'd like to have it for the next versions that I'm going to release soon).

**Comment by fabpot** [comment]
Thank you @seb-jean.

**Comment by seb-jean** [comment]
Thanks to the people who helped me build this PR.

**Pull request merged**